### PR TITLE
Optional keepalive to backend

### PR DESCRIPTION
**Pull request by ferd**
This is a newer copy of https://github.com/heroku/vegur/pull/141, rebased for current support.

The current implementation is still hacky as hell (using process
dictionaries), but lets us do some preliminary testing and benchmarking
regarding keeping communication open to backends.

The implementation chosen has the following properties:

- use the return value `{keepalive, {default, {IP, Port}}}` instead of just
  `{IP, Port}` when returning from `service_backend` callback.
  to know whether to turn on keepalive to the backend.
- The backend connection will be reused *only if the callback module
  implementing vegur_interface returns the tuple above*. Any
  other cases and the backend connection will be closed on the follow-up
  request.
- If the prior connection didn't use `keepalive` as a tuple but the new one
does, the previous connection will not be reused.
-  **Preventing someone from entering multiple requests over the
  connection addressed to multiple applications and breaking request
  routing or finding security exploits is not handled here.**
- The implementation relies on process dictionaries and tries to clean
  up after itself. It otherwise relies on connections to the front-end
  terminating to clean up resources.
  This means there are no connection pools, and none of the backend
  connections can last longer than the front-end connection that caused
  it to be established in the first place.
- Statistics about the request size and bytes transfered are maintained
  by storing a 'current offset' at the beginning of a session reuse.
  Doing it that way ensures that cleanup and reporting operations at the
  end of the prior request operate properly, but that all new usage
  does not end up breaking.
- The request log for the client gets reset; A default connect time of
  0 is set to be in place automatically to ensure the entry is neither
  missing nor to show it has actually taken any time. Other log
  structures for reports are external to the client and should be
  discarded automatically the way they were before.
- Tests have been written for most regular cases, and ensures that
  stream transfers (chunked, large bodies, close-delimited responses,
  upgraded connections) also interact well with keepalive.
- There is no keepalive connection limit enforced to the back-end. The
  assumption is that this can be carried on by the front-end side of
  things given the current implementation.
- Vegur should now send the connection:keep-alive header explicitly to a
  backend when the option is possible.

The following behaviour is specified for various combinations of client
requests' connection headers and backend response's connection headers,
given vegur already supported client keepalive:

   Option   | Connection | Response  |          Result
------------|------------|-----------|----------------------------
 close      | close      | close     | all connections are closed
 close      | close      | keepalive | all connections are closed
 close      | keepalive  | close     | only client connection is open
 close      | keepalive  | keepalive | only client connection is open
 keepalive  | close      | close     | all connections are closed
 keepalive  | close      | keepalive | all connections are closed
 keepalive  | keepalive  | close     | only client connection is open
 keepalive  | keepalive  | keepalive | all connections are open

An unspecified header value is assumed to be keep-alive, as per HTTP
specification. Upgrade requests can only be terminated by all
connections being closed as the protocol is no longer HTTP.
Close-delimited responses (without content-length) also forcefully
close all connections.

### CODE REVIEW

**Comment by ferd** [comment]
Something to consider and look at for this branch is that large body tests appear to be really really flaky somehow.

**Comment by ferd** [comment]
Related bug report sent to me via e-mail on the earlier prototype:

> Here at _____, we're attempting to productionalize a new metrics backend on top of _____. As luck would have it, this of course means that we're using the most experimental stuff [...], including the keepalive branch of Vegur.
>
> It seems as though we may have uncovered a bug, though I can't reliably reproduce it, and I'm wondering if you might have some ideas. We seem to get some percentage of requests failing with H25 due to content-lengths not being equal. We're not seeing many *other* errors. We've gotten 17 H18s and H99s in the same time period, and no others reported in Splunk.
>
> Given the nature of the commit message for the commit introducing keepalives for the Vegur branch we're on (https://github.com/heroku/vegur/commit/d32a0e61040cb6b6bd39f81f3b8063adadff4e94), I'm *hoping* you might have an idea, or be able to quickly discover a (probably) subtle bug that might cause this behaviour?
>
> I certainly don't want to place the blame on Vegur, but given that I've only seen matches in terms of content-length to body length on responses I've hand verified, and the maturity of the keepalive code, I'm thinking it's likely some nuance of the reconnect logic, a subtle bug in parsing or some other edge case thing.

That stuff is definitely possible and I'll have to look into it regarding test flakiness.

**Comment by ferd** [comment]
So the issue in `large_chunked_request_response_interrupt` is likely nothing:
```
=INFO REPORT==== 13-Jul-2016::20:07:11 ===

mod=vegur_proxy at=check message=downstream_buffer_full
```
This is probably a bit of nondeterminism and timing-dependent things and would be unrelated with the bug report.

**Comment by ferd** [comment]
Added a commit to allow proper replacement of keepalive connections.

In cases where the caller wants to do keepalive to the back-end with
limits on how many connections may be used (or may want to force a
replacement node to take the follow-up requests), the former semantics
only allowed to declare a continuing keepalive connection or a
non-keepalive one.

This yielded a weird case where if we have, say, 15 requests over a
client connection and want to send batches of 10 over keep-alive to a
back-end, the pattern required to do:

keepalive: 1..10
non-keepalive: 11
keepalive : 12..15

So instead of a more optimal path, the current API mandated the use of
an extra connection every time the limit was attained to avoid forcing a
keepalive conn.

This commit includes the option of specifying keepalive as either:

- {keepalive, {default, Backend}}
- {keepalive, {new, Backend}}

The usage of `new` here mandates forced closure of the backend currently
connected to, and calls for opening a new connection in its place that
will be ready for keepalive.

This allows proper support for a sequence being:

keepalive: 1..10
keepalive: 11..15

With only two connections.

A test has been added for the handling there.

File: src/vegur_client.erl [hunk 1]
```diff
@@ -841,6 +856,12 @@ set_stats(Client=#client{bytes_sent=BytesSent, bytes_recv=BytesRecv,
     Client#client{bytes_sent=Sent, bytes_recv=Recv}.
 
 %% @private
```
Comment (ypaq) [inline src/vegur_client.erl]: This is exported. Should it have a spec?

Comment (ferd) [inline src/vegur_client.erl]: It's an internal export, but sure, I can add it.

File: test/vegur_proxy_SUITE.erl [hunk 2]
```diff
@@ -112,6 +113,16 @@ end_per_group(_, Config) ->
     meck:unload(),
     Config.
 
+init_per_testcase(request_keepalive_statistics, Config) ->
+    %% force jitter on some operation to ensure stats vary between reqs
+    meck:new(vegur_headers, [passthrough]),
+    meck:expect(vegur_headers, request_headers,
+                fun(A,B,C) ->
+                        timer:sleep(random:uniform(20)),
```
Comment (ypaq) [inline test/vegur_proxy_SUITE.erl]: `rand`?

Comment (ferd) [inline test/vegur_proxy_SUITE.erl]: Will fix.

Comment (ferd) [inline test/vegur_proxy_SUITE.erl]: God damn, it is not 19-ready. That increases the scope a bit

**Comment by ypaq** [comment]
LGTM, as you mentioned the implementation using the process dictionary seems hacky. The test suite is really nice and comprehensive though.
